### DISCUSSION TITLE
How to generate documentation for Python-side of library?

### DISCUSSION THREAD

**Discussion by hwittenborn** [PyO3/maturin#1039] [Q&A] [answered]
I'm not sure if this is right place to ask, but I'm kind of lost rn and I was hoping someone could help out.

Anyway, so my library is exposed via both a Rust and Python library, and the Rust side for documentation was easy peasy, just publish to crates.io and their systems handled the rest. But I'm kind of lost at how I should generate my docs for the Python side of things. I've loosely heard of Read the Docs, and I tried dabbling with both Sphinx and mkDocs earlier, but I didn't get much of anywhere.

Is there any known system that I can use to generate my docs automatically in a fashing similar to how `cargo doc` does it? Publishing is another story, I honestly just want a way to get my docs *generated*, publishing and all that I should be able to handle just fine.

If it helps at all you can find my current project at <https://github.com/makedeb/makedeb-srcinfo>. Thanks a ton for this and PyO3 btw too :), it made developing a library for both ecosystems extremely easy, I don't think I put more than 30 minutes (if that) on getting the Python side of my library to just work flawlessly.

**Comment by messense** [comment]
Have you tried https://maturin.rs/sphinx.html?

**Reply by messense** [reply]
Here is an example project with both readthedocs and netlify integration: https://github.com/messense/rjieba-py

* https://rjieba-py.readthedocs.io/en/latest/
* https://rjieba-py.netlify.app/

**Reply by hwittenborn** [reply]
> Have you tried https://maturin.rs/sphinx.html?

I took a brief look at that earlier, I didn't get much of anywhere with it; I just did another more thorough read but still got stuck on what I should be doing rn.

From what I see that article mostly talks about setting up the build environment, which I was pretty much able to do fine. My issue was just the automatic doc generation with things like classes and functions being automatically documented, which that article doesn't seem to cover.

**Reply by hwittenborn** [reply]
Oh, didn't see that last comment :P checking it out rn

**Reply by messense** [reply]
It's not very much different from pure Python project, just follow the [Sphinx quickstart](https://www.sphinx-doc.org/en/master/usage/quickstart.html) guide then setup the build environment.

To generate API docs automatically, you can use Sphinx [autodoc](https://www.sphinx-doc.org/en/master/usage/quickstart.html#autodoc) or [autoapi](https://sphinx-autoapi.readthedocs.io/en/latest/) extension.

**Comment by hwittenborn** [comment, marked as answer]
I took another shot at this today after a bit of it not working, and it's just working now, no clue why :p. Anyway, looks like I'm set now, thanks a ton!
